AEP: VPA native sidecar support (#8905)
kubernetes/autoscaler: vertical-pod-autoscaler/enhancements/8905-native-sidecar-support/README.md

Reviewer: I think this line duplicates the one in the first bullet point, i.e. this: "Admission applies recommendations to native sidecars"

Reviewer: ```suggestion
by passing `--feature-gates=NativeSidecar=true` to all VPA components.
```

Reviewer 2: Do we have an official supported matrix for VPA and k8s? I mean, I probably wouldn't go further down than the [official skew policy](https://kubernetes.io/releases/version-skew-policy/). So in this particular case you can count on the feature being available. 

In general, we don't hard-code any version dependencies in the code, we always rely on feature availability. So you should probably change that statement accordingly, to make it explicit.

Reviewer 3: We have this: https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/docs/installation.md#compatibility

I've never liked it and I'm sure it's wrong.

I assume what we should do is have something similar to https://github.com/kubernetes/client-go#compatibility-matrix ?

Reviewer 2: Honestly, I'll leave the decision to you. I only care we have something that we can reference to our users.

Reviewer 3: Cool thanks. I've taken an action item to improve out compatibility table.

Reviewer 2: Suggesting, to avoid confusion, I'd call them `ContainerTypeNativeSidecar`.

Reviewer 2: My only question is about the difference in when the native sidecars are started vs regular containers. The former, being defined in the `initContainers` section benefit from early start, does that somehow affect VPA? Can we explicitly write it down here.

Reviewer 2: The other question stems from this section of the blog introducing native sidecards: https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/#resource-sharing-within-containers Does it impact any of the VPA changes?

Addresses [issue #7229](https://github.com/kubernetes/autoscaler/issues/7229)

### Goals

- Allow VPA Recommender to generate resource recommendations for native sidecar containers.
- Ensure VPA Updater and Admission Controller can apply recommendations to native sidecar containers.

### Non-Goals

- Support for sidecar containers in Kubernetes versions older than 1.28.

## Proposal

The proposal is to introduce a new feature gate `NativeSidecar` in VPA. When enabled, VPA components will recognize and handle native sidecar containers.

## Design Details

### Recommendations

The Recommender component identifies native sidecar containers by examining init containers with `restartPolicy: Always` in the `SpecClient`. These containers are assigned the `ContainerTypeInitSidecar` type.

When the `NativeSidecar` feature gate is enabled, the `ClusterFeeder` processes native sidecars similarly to standard containers:

> - They are stored in a separate `InitSidecarsContainers` map in the pod state

Author: for updates there is an extra array in the response from GetContainersResourcesForPod. 

so we could technically keep the feeder logic in  a single map if you'd like. I just thought thought it was clearer for readability with the feature flag at the main function too. let me know and i can update it.

# KEP-8905: Native Sidecar Support

<!-- toc -->
- [Summary](#summary)
   - [Goals](#goals)
   - [Non-Goals](#non-goals)
- [Proposal](#proposal)
- [Design Details](#design-details)
   - [Recommendations](#recommendations)
   - [Update / Admission](#update--admission)
   - [Test Plan](#test-plan)
   - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
- [Implementation History](#implementation-history)
- [Alternatives](#alternatives)
<!-- /toc -->

## Summary

This proposal adds support for native sidecar containers (init containers with `restartPolicy: Always`) in Vertical Pod Autoscaler. Standard init containers will continue to be ignored.

> Kubernetes 1.28 introduced native sidecar containers. These are init containers that start before the main containers and continue running during the lifecycle of the Pod. VPA currently supports standard containers and regular init containers, but it should also support recommending resources for these new native sidecar containers to ensure they are right-sized.

Author: yeah, will rephrase the paragraph about supporting it to ignore. the code did historically have them added to another array but for the purposes of VPA it doesn't make recommendations or act on them so will just say ignore.

Addresses [issue #7229](https://github.com/kubernetes/autoscaler/issues/7229)

### Goals

- Allow VPA Recommender to generate resource recommendations for native sidecar containers.
- Ensure VPA Updater and Admission Controller can apply recommendations to native sidecar containers.

### Non-Goals

- Support for sidecar containers in Kubernetes versions older than 1.28.

> ## Proposal

Author: > * to make sure that [capping recommendations according to a defined PodLimit](https://github.com/kubernetes/autoscaler/blob/0d95f31683fedaf42b9999e1f03c27689fba7fb9/vertical-pod-autoscaler/pkg/utils/vpa/capping.go#L367) still works

good call out, I updated it to use init containers as well. i think [this](https://github.com/kubernetes/autoscaler/pull/8498/changes#diff-71dc66606ae3da94d6f0b8978943b06cb5b68fa74900e579727400e1a8743934L35) was the main function that needed to be updated for it to work. 

> * to allow (optional?) source recommendations for native Sidecars, similar to what you can achieve today for non-native Sidecars by executing the VPA webhook last and with `reinvocationPolicy: ifNecessary`

I'm not sure we can do anything differently here, i also think that in place should help with this?

The proposal is to extend VPA to support native sidecar containers. This involves updating the Recommender to generate recommendations for these containers, and updating the Updater and Admission Controller to apply these recommendations.

This functionality will be guarded by a new feature gate `NativeSidecar` to allow for safe rollout.

> ## Design Details

Reviewer 4: > E.g., you can ignore containers in the VPA spec by specifying the containerresourcepolicy indexed by containerName, and setting mode to Off. I don't think it makes sense to make a new API type specifically for sidecar containers for this.

Why not? Since container names are unique, the existing `containerResourcePolicy` should be enough to target them without needing any API changes.

### Recommendations

The Recommender component identifies native sidecar containers by examining init containers with `restartPolicy: Always` in the `SpecClient`. These containers are assigned the `ContainerTypeInitSidecar` type.

When the `NativeSidecar` feature gate is enabled, the `ClusterFeeder` processes native sidecars similarly to standard containers:
- They are stored in a separate `InitSidecarsContainers` map in the pod state
- Resource usage samples are collected and aggregated for recommendations
- Recommendations are generated using the same logic as standard containers
```go
for _, container := range pod.Containers {
  if err := feeder.clusterState.AddOrUpdateContainer(container.ID, container.Request, container.ContainerType); err != nil {

>     klog.V(0).InfoS("Failed to add container", "container", container.ID, "error", err)

Reviewer 4: Yes please ( I know this is not your "responsibility" but it should be corrected ).

  }
}
for _, initContainer := range pod.InitContainers {
  if features.Enabled(features.NativeSidecar) && initContainer.ContainerType == model.ContainerTypeInitSidecar {
    if err := feeder.clusterState.AddOrUpdateContainer(initContainer.ID, initContainer.Request, initContainer.ContainerType); err != nil {

>       klog.V(0).InfoS("Failed to add initContainer", "container", initContainer.ID, "error", err)

Reviewer 4: ```suggestion
      klog.V(4).ErrorS(err, "Failed to add initContainer", "container", initContainer.ID)
```

```

The patch generation logic is updated to target `/spec/initContainers` for native sidecar containers, while normal containers will continue to update `/spec/containers`.

### Test Plan

The following test scenarios will be added to e2e tests.

- Admission applies recommendations to native sidecars.
- Updater will update sidecar container resources in-place or evict.
- Admission will patch sidecar container resources.
- When the feature gate `NativeSidecar` is false VPA components will not modify native sidecars.
- Verify that regular init containers (without `restartPolicy: Always`) are not treated as sidecars and do not receive recommendations intended for sidecars.

### Upgrade / Downgrade Strategy

#### Upgrade

On upgrade of the VPA to 1.6.0 (tentative release version), nothing will change,
VPAs will continue to work as before.

Users can use the new `NativeSidecar` by enabling the alpha Feature Gate (which defaults to disabled)
by passing `--feature-gates=NativeSidecar=true` to the VPA components.

#### Downgrade

On downgrade of VPA from 1.6.0 (tentative release version), nothing will change.

> VPAs will continue to work as previously. Checkpoints may contain sidecar resource information until updated, but updater and admission will not modify sidecar resources.

Reviewer 4: Correct.

### Update / Admission

`GetContainersResourcesForPod` has been updated to retrieve recommendations for native sidecars, Updater and Admission Controller both use this method so the logic is fairly contained.

```go
GetContainersResourcesForPod(pod *core.Pod, vpa *vpa_types.VerticalPodAutoscaler) ([]vpa_api_util.ContainerResources /*containers*/, []vpa_api_util.ContainerResources /*native sidecars*/, vpa_api_util.ContainerToAnnotationsMap, error)
```

The patch generation logic is updated to target `/spec/initContainers` for native sidecar containers, while normal containers will continue to update `/spec/containers`.

### Test Plan

The following test scenarios will be added to e2e tests.

- Admission applies recommendations to native sidecars.
- Updater will update sidecar container resources in-place or evict.
- Admission will patch sidecar container resources.
- When the feature gate `NativeSidecar` is false VPA components will not modify native sidecars.
- Verify that regular init containers (without `restartPolicy: Always`) are not treated as sidecars and do not receive recommendations intended for sidecars.

### Upgrade / Downgrade Strategy

#### Upgrade

> Upgrading to VPA 1.7.0 (tentative release version) does not change behavior by default. Existing VPAs continue to work as before.

Reviewer 3: ```suggestion
Upgrading to VPA 1.80 (tentative release version) does not change behavior by default. Existing VPAs continue to work as before.
```

🤞 Hopefully!

Reviewer 5: 1.8.0* :-)

To use native sidecar support, enable the (alpha) `NativeSidecar` feature gate (disabled by default) by passing `--feature-gates=NativeSidecar=true` to all VPA components.
When enabled, the Recommender will generate recommendations for native sidecars, and the Updater and Admission Controller will apply them.

#### Downgrade

> If the `NativeSidecar` feature gate was never enabled, downgrading from VPA 1.7.0 (tentative release version) has no behavioral impact.

Reviewer 3: ```suggestion
If the `NativeSidecar` feature gate was never enabled, downgrading from VPA 1.8.0 (tentative release version) has no behavioral impact.
```

## Design Details

### Recommendations

The Recommender component identifies native sidecar containers by examining init containers with `restartPolicy: Always` in the `SpecClient`. These containers are assigned the `ContainerTypeInitSidecar` type.

When the `NativeSidecar` feature gate is enabled, the `ClusterFeeder` processes native sidecars similarly to standard containers:
- They are stored in a separate `InitSidecarsContainers` map in the pod state
- Resource usage samples are collected and aggregated for recommendations
- Recommendations are generated using the same logic as standard containers

The VPA custom resource definition remains unchanged. Native sidecar recommendations are included in the `containerRecommendations` array alongside standard container recommendations, using the unique container name to identify them.

### Update / Admission

Both Updater and Admission Controller components retrieve recommendations for init containers (including native sidecars) separately from standard containers using `GetContainersResourcesForPod`.

> The patch generation logic is updated to target `/spec/initContainers` for native sidecar containers, while normal containers will continue to update `/spec/containers`.

Reviewer 4: Personally, I would prefer to include the updated code snippets in the AEP so users can clearly understand what this change looks like.
